ragna/assistants/_google.py

```diff
@@ -0,0 +1,117 @@
+from typing import AsyncIterator
+
+from ragna._compat import anext
+from ragna.core import PackageRequirement, Requirement, Source
+
+from ._api import ApiAssistant
+
+
+class AsyncIteratorReader:
+    def __init__(self, ait: AsyncIterator[bytes]) -> None:
+        self._ait = ait
+
+    async def read(self, n: int) -> bytes:
+        if n == 0:
+            return b""
+        return await anext(self._ait, b"")  # type: ignore[call-arg]
+
+
+class GoogleApiAssistant(ApiAssistant):
+    _API_KEY_ENV_VAR = "GOOGLE_API_KEY"
+    _MODEL: str
+    _CONTEXT_SIZE: int
+
+    @classmethod
+    def requirements(cls) -> list[Requirement]:
+        return [
+            *super().requirements(),
+            PackageRequirement("ijson"),
+        ]
+
+    @classmethod
+    def display_name(cls) -> str:
+        return f"Google/{cls._MODEL}"
+
+    @property
+    def max_input_size(self) -> int:
+        return self._CONTEXT_SIZE
+
+    def _instructize_prompt(self, prompt: str, sources: list[Source]) -> str:
+        # https://ai.google.dev/docs/prompt_best_practices#add-contextual-information
+        return "\n".join(
+            [
+                "Answer the prompt using only the pieces of context below.",
+                "If you don't know the answer, just say so. Don't try to make up additional context.",
+                f"Prompt: {prompt}",
+                *[f"\n{source.content}" for source in sources],
+            ]
+        )
+
+    async def _call_api(
+        self, prompt: str, sources: list[Source], *, max_new_tokens: int
+    ) -> AsyncIterator[str]:
+        import ijson
+
+        async with self._client.stream(
+            "POST",
+            f"https://generativelanguage.googleapis.com/v1beta/models/{self._MODEL}:streamGenerateContent",
+            params={"key": self._api_key},
+            headers={"Content-Type": "application/json"},
+            json={
+                "contents": [
+                    {"parts": [{"text": self._instructize_prompt(prompt, sources)}]}
+                ],
+                # https://ai.google.dev/docs/safety_setting_gemini
+                "safetySettings": [
+                    {"category": f"HARM_CATEGORY_{category}", "threshold": "BLOCK_NONE"}
+                    for category in [
+                        "HARASSMENT",
+                        "HATE_SPEECH",
+                        "SEXUALLY_EXPLICIT",
+                        "DANGEROUS_CONTENT",
+                    ]
+                ],
+                # https://ai.google.dev/tutorials/rest_quickstart#configuration
+                "generationConfig": {
+                    "temperature": 0.0,
```

Comment:
It is the other way around: we want to hardcode `0.0` here, because that means determinism. If we learned one thing from trying to bring RAG to businesses is that they want to get exactly the same answer if they ask the same question twice. Of course we can't guarantee it since we don't control the model, but we can do our best to at least avoid sampling during generation.

This is the same for all other assistants that we currently have

https://github.com/Quansight/ragna/blob/1b53e62cb873ba16968b7c25a81baff0947d705c/ragna/assistants/_anthropic.py#L52
https://github.com/Quansight/ragna/blob/1b53e62cb873ba16968b7c25a81baff0947d705c/ragna/assistants/_mosaicml.py#L43
https://github.com/Quansight/ragna/blob/1b53e62cb873ba16968b7c25a81baff0947d705c/ragna/assistants/_openai.py#L58